lua: add three new methods to import the private key and to encrypt/decrypt the data

agrawroh opened the pull request:
### Background

Today we already have a method to load the public key and use it to verify the signatures. This PR adds three new methods to extend the crypto functionality by letting users import the private key and then use these public & private keys to encrypt/decrypt the incoming request data.

### Changes

This PR adds the following three new methods to LUA filter:
1. **importPrivateKey()** - This method can be used to load the private key similar to what we have for loading the public key i.e. `importPublicKey()`. 
2. **encrypt()** - This method can be used to encrypt a given plaintext message using the public key.
3. **decrypt()** - This method can be used to decrypt a given cipher text message using the private key.
---
**Commit Message:** add three new methods to import the private key and to encrypt/decrypt the data.
**Additional Description:** See Background Section.
**Risk Level:** Low
**Testing:** Unit Tests
**Docs Changes:** Added the description of all the three new methods which we are adding in LUA. 
**Release Notes:** Added
**Platform Specific Features:** N/A

**Signed-off-by:** Rohit Agrawal <rohit.agrawal@databricks.com>

mattklein123 commented:
I think the coverage issue is legitimate. Can you look at that?

/wait

mattklein123 commented:
clang tidy issue looks legit.

/wait

agrawroh commented:
@mattklein123 Thanks, I'll resolve the lint issues but I couldn't get coverage beyond 83.xx% as I can't figure out a way to mock the OpenSSL methods to initialize the context. Any suggestions on what could be done?

mattklein123 commented:
> Thanks, I'll resolve the lint issues but I couldn't get coverage beyond 83.xx% as I can't figure out a way to mock the OpenSSL methods to initialize the context. Any suggestions on what could be done?

Sorry no. If you need to reduce coverage so be it.

/wait

RyanTheOptimist commented:
> @mattklein123 Thanks, I'll resolve the lint issues but I couldn't get coverage beyond 83.xx% as I can't figure out a way to mock the OpenSSL methods to initialize the context. Any suggestions on what could be done?

Is there a link to a coverage run I could look at? (I don't see one in the current list of CI jobs)

agrawroh commented:
@RyanTheOptimist Here is the one from last run:
https://dev.azure.com/cncf/envoy/_build/results?buildId=143761&view=logs&j=bbe4b42d-86e6-5e9c-8a0b-fea01d818a24&t=299ff00a-2cef-5552-88aa-ea9cf69316f0

RyanTheOptimist commented:
> @RyanTheOptimist Here is the one from last run: https://dev.azure.com/cncf/envoy/_build/results?buildId=143761&view=logs&j=bbe4b42d-86e6-5e9c-8a0b-fea01d818a24&t=299ff00a-2cef-5552-88aa-ea9cf69316f0

Thanks! looks like one thing we could do to get a bit more coverage would be to test `UtilityImpl::getHashFunction()` as it looks like there is some low hanging fruit there: 
https://storage.googleapis.com/envoy-pr/ccfbd87/coverage/source/common/crypto/utility_impl.cc.gcov.html

Looks like we could also cover the setters in 
https://storage.googleapis.com/envoy-pr/ccfbd87/coverage/source/common/crypto/crypto_impl.cc.gcov.html

Could you add some basic/trivial tests for those case and see what that does to the coverage numbers?

agrawroh commented:
@RyanTheOptimist Sure, given `getHashFunction()` is already checked in code, you want me to include these new tests in this PR or is it okay if I create a followup with additional tests and restore the numbers based on what we achieve with the new additional tests?

wbpcode commented on source/common/crypto/utility_impl.cc:
```diff
@@ -35,6 +35,86 @@ std::vector<uint8_t> UtilityImpl::getSha256Hmac(const std::vector<uint8_t>& key,
+  }
+
+  // Step 2: initialize EVP_PKEY_CTX
+  EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new(pkey, nullptr);
```
May be we can create a simple struct which will call the `EVP_PKEY_CTX_free(ctx)` automatically when it's destructed.
Then we needn't call `EVP_PKEY_CTX_free(ctx)` everywhere.

RyanTheOptimist commented:
> @RyanTheOptimist Sure, given `getHashFunction()` is already checked in code, you want me to include these new tests in this PR or is it okay if I create a followup with additional tests and restore the coverage numbers based on what we achieve with the new additional tests?

Please add them to this PR so we can get a clear picture of what they do to the coverage numbers.

RyanTheOptimist commented:
/wait

agrawroh commented:
> > @RyanTheOptimist Sure, given `getHashFunction()` is already checked in code, you want me to include these new tests in this PR or is it okay if I create a followup with additional tests and restore the coverage numbers based on what we achieve with the new additional tests?
> 
> Please add them to this PR so we can get a clear picture of what they do to the coverage numbers.

@RyanTheOptimist I added additional test cases as per your suggestion and also incorporated the wrapper changes  suggested by @wbpcode. Could you please take another look?

agrawroh commented:
/retest

agrawroh commented:
/retest

agrawroh commented:
/retest

RyanTheOptimist commented:
> > > @RyanTheOptimist Sure, given `getHashFunction()` is already checked in code, you want me to include these new tests in this PR or is it okay if I create a followup with additional tests and restore the coverage numbers based on what we achieve with the new additional tests?
> > 
> > 
> > Please add them to this PR so we can get a clear picture of what they do to the coverage numbers.
> 
> @RyanTheOptimist I added additional test cases as per your suggestion and also incorporated the wrapper changes suggested by @wbpcode. Could you please take another look?
> 
> [New Coverage](https://storage.googleapis.com/envoy-pr/4c37795/coverage/source/common/crypto/index.html) is around 90.6%.

Great! Looks like this PR no longer needs to lower coverage threshold. Thanks for doing this.

agrawroh commented:
/retest

RyanTheOptimist commented:
/assign-from @envoyproxy/envoy-maintainers

RyanTheOptimist commented:
/assign @wbpcode 
Who has some context about this

ggreenway commented:
Just one high-level comment: I believe RSA is the only private-key type that supports encrypt/decrypt. Using other key types for these operations will result in an error. This should be documented and tested.

Also, this code sets up encrypt/decrypt only using the default parameters in boringssl. This may or may not be the desirable settings, depending on usage. For example, you may want to call `EVP_PKEY_CTX_set_rsa_padding` to specify a specific padding mode. The boringssl documentation says

```
// The |padding| argument must be one of the |RSA_*_PADDING| values. If in
// doubt, use |RSA_PKCS1_OAEP_PADDING| for new protocols but
// |RSA_PKCS1_PADDING| is most common.
```

The default mode is `RSA_PKCS1_PADDING`. All of this should probably be documented, or at least link back to the boringssl docs for the APIs you're using.

KBaichoo commented:
PTAL @wbpcode 🌷

KBaichoo commented:
can you take a look @wbpcode ? Thank you

wbpcode commented:
I think the last comment from @ggreenway is not be addressed?

/wait-any
